axlearn/cloud/gcp/dataflow_inference_custom.py

```diff
@@ -0,0 +1,205 @@
+        --trainer_dir='gs://.../checkpoints/step_xxx' \
+        '"
+
+"""
+
+
+import copy
+import logging
+import warnings
+from typing import Any, Dict, Optional, Sequence
+
+import apache_beam as beam
+import jax
+from absl import app, flags
+from absl.flags import argparse_flags
+from apache_beam.ml.inference.base import ModelHandler, PredictionResult, RunInference
+from apache_beam.options.pipeline_options import PipelineOptions
+
+import axlearn.common.input_fake as input_fake
+import axlearn.common.launch_trainer as trainer_utils
+from axlearn.common.inference import InferenceRunner, MethodRunner
+from axlearn.common.utils import NestedTensor
+
+warnings.filterwarnings("ignore")
+
+
+class CustomModelHandler(ModelHandler[Dict, PredictionResult, Any]):
+    """Defines how to load a model and run inference"""
+
+    def __init__(self, flag_dict):
+        self._flag_dict = flag_dict
+
+    def _flag_values_from_dict(self, flag_values: dict) -> flags.FlagValues:
+        # Avoid mutating global FLAGS.
+        fv = copy.deepcopy(flags.FLAGS)
+        for k, v in flag_values.items():
+            try:
+                fv.set_default(k, v)
+            except flags._exceptions.UnrecognizedFlagError:
+                # Ignore unrecognized flags from other modules
+                pass
+        fv.mark_as_parsed()
+        return fv
+
+    def load_model(self) -> MethodRunner:
+        """Loads a pre-trained model in the desired type (MethodRunner in this case).
+        Reference: https://github.com/apple/axlearn/blob/main/axlearn/common/inference.py#L54
+
+        Returns:
+          An instance of MethodRunner.
+        """
+        # construct absl FlagValues from dict
+        flag_values = self._flag_values_from_dict(self._flag_dict)
+
+        module_config = trainer_utils.get_trainer_config(flag_values=flag_values)
+
+        # get InferenceRunner Config from Trainer Config and instantiate InferenceRunner
+        inference_runner_cfg = InferenceRunner.config_from_trainer(module_config)
+        inference_runner_cfg.init_state_builder.set(dir=flag_values.trainer_dir)
+        inference_runner = InferenceRunner(cfg=inference_runner_cfg, parent=None)
+
+        # create Method Runner only once
+        method_runner = inference_runner.create_method_runner(
+            method="predict", prng_key=jax.random.PRNGKey(1)
+        )
+        return method_runner
+
+    def run_inference(
+        self,
+        batch: Sequence[NestedTensor],
+        model: MethodRunner,
+        inference_args: Optional[Dict[str, Any]] = None,
+    ):
+        """Runs inferences on a batch of NestedTensors.
+        NestedTensor: https://github.com/apple/axlearn/blob/main/axlearn/common/utils.py#L56
+
+        Args:
+          batch: A sequence of examples as NestedTensors.
+          model: An instance of a MethodRunner.
+          inference_args: Any additional arguments for an inference.
+
+        Returns:
+          A list of type MethodRunner.Output.
```

Comment:
```suggestion
        Args:
            batch: A sequence of examples as NestedTensors.
            model: An instance of a MethodRunner.
            inference_kwargs: Optional additional keyword arguments for inference.

        Returns:
            A list of method runner outputs.
```

In general, we do not need to spell out the types in the docstring, since they are already in the fn signature.
Also, it seems that we do not make use of `inference_kwargs`, is that intended?

Reply:
Yes, so that end users can add more args if they want to.